### ISSUE TITLE
[Fatal Error] LEAK: ByteBuf.release() was not called before it's garbage-collected

### ISSUE THREAD

**Issue by Joshuaalbert** [neo4j/neo4j#13207]
We just had a system crash due to this one. The logs imply there is a memory handling error at the heart of it.

# Bug Report
```bash
2023-05-30 14:03:44.791+0000 ERROR LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records:
Created at:
        io.netty.buffer.AbstractByteBufAllocator.compositeDirectBuffer(AbstractByteBufAllocator.java:224)
        io.netty.buffer.AbstractByteBufAllocator.compositeBuffer(AbstractByteBufAllocator.java:202)
        io.netty.handler.codec.MessageAggregator.decode(MessageAggregator.java:271)
        io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
        io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327)
        io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299)
        io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        io.netty.handler.codec.ByteToMessageDecoder.handlerRemoved(ByteToMessageDecoder.java:255)
        io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:517)
        io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:449)
        io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:279)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        org.neo4j.bolt.transport.pipeline.AuthenticationTimeoutHandler.channelRead(AuthenticationTimeoutHandler.java:58)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800)
        io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
        io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
        io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        java.base/java.lang.Thread.run(Thread.java:829)
```

# My Environment
Python Version: 3.10.8
Driver Version: `5.8.0`
Server Version and Edition: `neo4j-community-4.4.9-unix`
Operating System: Ubuntu

**Comment by gjmwoods** [comment]
Hi @Joshuaalbert, many thanks for raising this issue. Since this is an error originate on the server we've transfered this over to the server repo.

As the error indicates, the logs show that some memory has been leaked somewhere during authentication on bolt connections. These leaks can be very tricky to isolate so it would be great if you could provide some additional context of the workload you were running at the time. E.g. which auth mechanism are you using (basic, bearer etc) and do you see lots of Authentication timeouts on the client side?

Are there any more logs available? This error usually serves as a warning that memory hasn't been collected properly but typically the server would crash on a different error (e.g. Out of memory)

**Comment by Joshuaalbert** [comment]
Hi @gjmwoods thanks for the reply. We're using bearer authentication. Regarding the workload, I think we're holding DB connections for fairly long (hours, held by services, and terminated at exit of service). We're creating sessions (via python driver with no `connection_acquisition_timeout`) ephemerally when needed for every transaction. I'll link @get-a-grep for answering what we see in the client/server logs.

**Comment by get-a-grep** [comment]
Hi @gjmwoods thanks for the response, as far as the the error logs go, there seem to be two relevant errors:

For the connection issue:
```
2023-05-30 14:03:20.708+0000 ERROR [o.n.b.t.p.ProtocolHandshaker] Fatal error occurred during protocol handshaking: [id: 0x43eca1f2, L:/<SERVER-IP>:7687 - R:/<CLIENT-IP>:<CLIENT-PORT>]
java.lang.NullPointerException: null
        at io.netty.buffer.Unpooled.copiedBuffer(Unpooled.java:363) ~[netty-buffer-4.1.75.Final.jar:4.1.75.Final]
        at org.neo4j.bolt.transport.DiscoveryResponseHandler.channelRead(DiscoveryResponseHandler.java:60) ~[neo4j-bolt-4.4.9.jar:4.4.9]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) [netty-codec-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327) [netty-codec-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299) [netty-codec-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.handlerRemoved(ByteToMessageDecoder.java:255) [netty-codec-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:517) [netty-codec-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:449) [netty-codec-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:279) [netty-codec-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at org.neo4j.bolt.transport.pipeline.AuthenticationTimeoutHandler.channelRead(AuthenticationTimeoutHandler.java:58) [neo4j-bolt-4.4.9.jar:4.4.9]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800) [netty-transport-classes-epoll-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480) [netty-transport-classes-epoll-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378) [netty-transport-classes-epoll-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [netty-common-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.75.Final.jar:4.1.75.Final]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.75.Final.jar:4.1.75.Final]
        at java.lang.Thread.run(Thread.java:829) [?:?]
```
For an IO issue:
```
2023-05-30 19:40:32.653+0000 ERROR [o.n.k.i.t.l.c.CheckPointerImpl] [neo4j/478c55af] Checkpoint failed
java.nio.file.AccessDeniedException: /data/transactions/neo4j/checkpoint.11
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:90) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116) ~[?:?]
        at sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:182) ~[?:?]
        at java.nio.channels.FileChannel.open(FileChannel.java:292) ~[?:?]
        at org.neo4j.io.fs.DefaultFileSystemAbstraction.open(DefaultFileSystemAbstraction.java:76) ~[neo4j-io-4.4.9.jar:4.4.9]
        at org.neo4j.io.fs.DefaultFileSystemAbstraction.write(DefaultFileSystemAbstraction.java:107) ~[neo4j-io-4.4.9.jar:4.4.9]
        at org.neo4j.io.fs.DefaultFileSystemAbstraction.write(DefaultFileSystemAbstraction.java:58) ~[neo4j-io-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.files.TransactionLogChannelAllocator.allocateFile(TransactionLogChannelAllocator.java:151) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.files.TransactionLogChannelAllocator.createLogChannel(TransactionLogChannelAllocator.java:64) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.checkpoint.DetachedCheckpointAppender.rotateChannel(DetachedCheckpointAppender.java:158) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.checkpoint.DetachedCheckpointAppender.rotate(DetachedCheckpointAppender.java:148) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.files.checkpoint.CheckpointLogFile.rotate(CheckpointLogFile.java:256) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.rotation.FileLogRotation.doRotate(FileLogRotation.java:147) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.rotation.FileLogRotation.locklessRotateLogIfNeeded(FileLogRotation.java:118) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.rotation.FileLogRotation.rotateLogIfNeeded(FileLogRotation.java:87) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.checkpoint.DetachedCheckpointAppender.checkPoint(DetachedCheckpointAppender.java:128) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.checkpoint.CheckPointerImpl.doCheckPoint(CheckPointerImpl.java:208) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.checkpoint.CheckPointerImpl.checkPointIfNeeded(CheckPointerImpl.java:170) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.transaction.log.checkpoint.CheckPointScheduler$1.run(CheckPointScheduler.java:67) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.scheduler.ScheduledJobHandle.lambda$new$0(ScheduledJobHandle.java:122) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.scheduler.ThreadPool.lambda$asCallable$1(ThreadPool.java:151) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at org.neo4j.kernel.impl.scheduler.ThreadPool.lambda$submit$0(ThreadPool.java:115) ~[neo4j-kernel-4.4.9.jar:4.4.9]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
```

The IO Issue crashes neo4j and is a bit nebulous as the access permissions are all accurately assigned. The only cause I could think of off the top of my head would be a lock that wasn't released on the checkpoint. If you'd like the full debug logs for that day, let me know how I should dump those on here (they're fairly large, don't know if there's a file size restriction).

**Comment by gjmwoods** [comment]
@Joshuaalbert @get-a-grep many thanks for the additional information. Seems like there's a few things things going on here and they might not all be related:

1) The memory leak `ERROR LEAK: ByteBuf.release()`
2) The `Fatal error occurred during protocol handshaking`
3) Checkpointing issue.

We've uncovered a bug for No 2 which since it also effects memory may also resolve No 1. We'll get working on the fix for that.

For No 3 I will ask my colleagues from the kernel team to take a look as it appears outside of the driver/bolt server.

**Comment by gjmwoods** [comment]
Kernel have said that No 3 should be something that a recovery/restart should resolve. If that fails some more investigations on the deployment machine to uncovery how the neo4j process loses access permissions on that file.

**Comment by Joshuaalbert** [comment]
Thanks @gjmwoods for the reply, good to know there is a potential fix for at least part of this underway. We do restart the service in these situations, and will continue to report here.

**Comment by gjmwoods** [comment]
@Joshuaalbert I have merged the fix for the `Fatal error occurred during protocol handshaking` issue and it should be available in the next 4.4 release which is scheduled for the end on June. I would suggest that the next steps are to try that version out and see if the leak issue reappears.

I didn't catch this earlier but also bearer authentication isn't supported in community edition.

**Comment by gjmwoods** [comment]
@Joshuaalbert Server version `4.4.22` has been release with a fix for this bug. I will close this issue for now. Please do reopen it if reappears. Thanks again for raising the issue.
